### ISSUE TITLE
HOWTO set up is_staff and is_superuser from Azure to App

### ISSUE THREAD

**Issue by ebreton** [snok/django-auth-adfs#219]
Hi folks,

Firstofall, thanks for the amazing lib. I find it clean and efficient. 🤩 

Documentation is great too, and I could setup the basics in Azure thanks to the explicit tutorial. Thanks a lot !

I wanted to go further, and I was looking for setting up user rights through `BOOLEAN_CLAIM_MAPPING`, and I am hitting the Azure wall of a-kilometer-high documentation. 

**Hence my question : could you add a section in the doc that describes the steps...**

1. to declare and set the attribute user_is_staff and user_is_superuser in Azure
1. add the claims in the token
1. (and get them, but it looks like this part is already working fine :)

**🧭 Status Point**

Here are my findings of the last 24 hours trying to make it work on my own (being completely newbie with Azure)

I started from the settings reference in the project doc, about `BOOLEAN_CLAIM_MAPPING`

> You can find the short name for the claims you configure in the ADFS management console underneath ADFS ➜ Service ➜ Claim Descriptions

I did not find such section in the Azure AD I created, neither in the Azure AD B2C :(

**📔 Regarding AD:**

The posts I have found redirect to the token configuration page, showing how to add [optional claims](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-optional-claims), but only existing ones apparently...

Therefore I searched how to define my claims, i.e. my new attributes on the users, but I only found an  azure feature [custom security attribute](https://docs.microsoft.com/en-us/azure/active-directory/fundamentals/custom-security-attributes-manage) which require a license and is still in PREVIEW.

It appears I am not on the good track

**📚 Therefre I looked at AD B2C:**

There, it looks simple to adjust the user attribute configuration (only availalbe on AD B2C though)

![attributes](https://user-images.githubusercontent.com/490665/152643736-4a624f6e-f569-41ea-afd6-f48865e496b9.png)

But... there is no "token configuration" page as mentioned in a lot of places. I found the reason in the [Microsoft doc ](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-optional-claims)

> The UI option Token configuration blade is not available for apps registered in an Azure AD B2C tenant which can be configured by modifying the application manifest. For more information see [Add claims and customize user input using custom policies in Azure Active Directory B2C](https://docs.microsoft.com/en-us/azure/active-directory-b2c/configure-user-input)

Regarding the claims token configuration, it appears that I have to use user flows or policies (https://docs.microsoft.com/en-us/azure/active-directory-b2c/configure-tokens?pivots=b2c-user-flow)

But that did not look like the way to go either

**🕵️‍♂️  Conclusion:**

I am pretty sure I am just getting lost in Azure labyrithn and missing an obvious configuration. 

Hence the my issue, and the suggestion about extending the Tutorial :)

If this process sounds trivial or is well-known to anyone, I am glad for any help

Cheers,
Emmanuel

**Comment by JonasKs** [comment]
Hi! I'm out skiing so I'll give you a little tip, and we can troubleshoot more another day. 


If you want to make a user super user based on a group, you can use this setting:

```py
'GROUP_TO_FLAG_MAPPING': {
        'is_staff': 'MyAdminGroup'
    },
```

**Comment by ebreton** [comment]
Thanks for the quick answer, on the skiing slopes ⛷️ 

I have setup Groups in Azure and the settings accordingly, but I still have something astray with the token

```
DEBUG 2022-02-05 16:19:42,583 django_auth_adfs The configured groups claim 'roles' was not found in the access token
DEBUG 2022-02-05 16:19:42,583 django_auth_adfs The configured group claim was not found in the access token
```

I will try again to look at the user flow configuration in the meantime :)

Cheers

**Comment by JonasKs** [comment]
😊

In Azure you have application roles and group claims. I recommend using application roles 😊 

https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-add-app-roles-in-azure-ad-apps

**Comment by ebreton** [comment]
Hi @JonasKs ,

**TLDR;**
I managed to setup the workaround through `groups` and app `roles` efficiently on Azure AD (not B2C). Thanks a lot for unblocking me 🥇 

That would probably a nice addition to the doc, I can make a PR if you want.

**My more detailed findings:**

1. **Azure AD B2C** is definitively a whole different beast to configure. 

The blocking issue I face is that you simply do not have the features available. Here is a screenshot of an app, from the `Registration Application` menu:

![App roles - B2C](https://user-images.githubusercontent.com/490665/152679250-10413b3e-938f-4131-a4a7-d93b56dc2b43.png)

Compared to what is in the Azure AD : 

![Capture d’écran 2022-02-06 à 12 46 08](https://user-images.githubusercontent.com/490665/152679358-74d820d1-3234-43ee-ba99-5d10d14abbb6.png)

No `Token configuration` entry, no `App role` entry in B2C

🆘   I put it aside for now, but if you have any experience with it, that will be very welcomed !

2. **Azure AD**

I face another issue with the `USERNAME_CLAIM`, which is `upn` for local members, but `email` for invited ones

🔬  What about supporting a list of `USERNAME_CLAIMS` where you could try the multiple keys ?

I am happy to make a PR if that makes sense for you

3. **Last but not least, user attributes...**

I am still lost on how you can define properties in Azure AD, and pass them to the app through claims. Any insight is welcomed again. I guess that remains the main topic of the discussion :)

Cheers,
Emmanuel

**Comment by JonasKs** [comment]
Hi, first of, Thank you for the detailed explanations. It is so much easier to help people who put an effort into explaining. 😊

I'm not on a PC this weekend, but I think #166 already solves your use case for usernames? Could you check that out? 

As for the rest, I'll read through again tomorrow and see if I have any input n

**Comment by ebreton** [comment]
My pleasure :)

I have tried the `GUEST_USERNAME_CLAIM` It looks like exactly what I need... but... the account I am testing sonehow has the same tenant id (how come????)... hence I am not satisfying all conditions, therefore I am still getting the `User claim's doesn't have the claim 'upn' in his claims`

I am not sure how it comes since I am using a professionnal AD with my personnal microsoft account, therefore not related to the AD before I invited myself. When double-checking in the AD, I can see that my personnal email is flagged as invited: `emmanuel@xxx.ch#EXT#@xxxx.onmicrosoft.com`

I tried another account (my github one). Following the sign-in process, I first got this error message:

```
AADSTS50020: User account 'manu@xxx.com' from identity provider 'live.com' 
does not exist in tenant 'Hymexia AD' and cannot access the application 
'xxxxxx'(Hymexia Fleet Manager) in that tenant. The account needs to be added 
as an external user in the tenant first. Sign out and sign in again with a different 
Azure Active Directory user account.
```

Therefore I invited this account, which looks as straightforward as it could be :)

![Capture d’écran 2022-02-07 à 18 25 40](https://user-images.githubusercontent.com/490665/152839871-4ccb1c9b-6627-4356-a9ec-10adc5dbbef5.png)
 
I go through the invitation process and hit the same error as before, because the tenant id is the same as the AD.

What could explain is that I had to kind of register into Hymexia AD during the the invitation process, including setting a password (which makes no sense since I have then logged with my github account)

End of the story: I am definitively lacking knowledge on Azure AD itself... 🙄
